### PR TITLE
remove the use of RTTI to tell shape types

### PR DESCRIPTION
**Pull request by pigpigyyy**
Add a small trick to remove the use of RTTI features to tell different shape types.

### CODE REVIEW

**Comment by ninnghazad** [comment]
I like the idea - however this makes ids somewhat unpredictable. dependent on order of first use or something like that. in many cases that doesn't matter - however if you were to serialize state, to files or network, that could become a problem - because it is not clearly visible to the user in what order the ids are generated or how to influence that order. maybe there is a way to force a fixed order. so no matter which type i use first, the order is static.

**Comment by pigpigyyy** [comment]
Yes, there is a way to force shape id generated in fixed order, just run the codes below before any shape creation.
```cpp
namespace pd = playrho::d2;
pd::ShapeType<pd::DiskShapeConf>(); // disk shape id 1
pd::ShapeType<pd::EdgeShapeConf>(); // edge shape id 2
pd::ShapeType<pd::ChainShapeConf>(); // chain shape id 3
pd::ShapeType<pd::PolygonShapeConf>(); // polygon shape id 4
pd::ShapeType<pd::MultiShapeConf>(); // multi shape id 5
```

File: PlayRho/Dynamics/World.cpp [hunk 1]
```diff
@@ -2111,6 +2111,7 @@ ContactCounter World::FindNewContacts()
     // to eliminate any node pairs that have the same body here before the key pairs are
     // sorted.
     for_each(cbegin(m_proxies), cend(m_proxies), [&](ProxyId pid) {
+        if (pid == DynamicTree::GetInvalidSize()) return;
```
Comment (louis-langholtz) [inline PlayRho/Dynamics/World.cpp:2114]: Thank you super much for recognizing this problem! Looks like you've made two different PRs which can address the fundamental problem you'd discovered: `m_proxies` can have invalid IDs in it before `FindNewContacts` is called.

I'm interested in unit test code that checks for this problem and am working on at least one check now for this. Sorry this wasn't caught already by a unit test. Glad at least it was caught by an assert.

Comment (louis-langholtz) [reply PlayRho/Dynamics/World.cpp:2114]: Thanks again for your help with this! I believe this is fixed now with pull request #341 that I've merged into master. If you can confirm, I'd very much appreciate it!

**Comment by louis-langholtz** [comment]
Crud! I hadn't meant to close this but only to delete my updates-20181223 branch. Hadn't realized till now that this PR is against the changes in that now deleted branch.

**Comment by louis-langholtz** [comment]
I've been thinking about this more. I really like the suggestion of removing the use of RTTI to tell shape types.

I recall reading somewhere previously about the technique @pigpigyyy uses. I'd thought I'd seen it in "Modern C++ Design" by Andrei Alexandrescu but I'm not finding it in there now. In any case, I dislike that this technique uses the incremented value of a static variable. That seems a bit like using a non-constant global variable.

An alternative technique is to use the address of a static function of a template class that's templated on the type. I believe using the address of a static variable of a template class could work too. By taking the address of the static function or static variable, there's no mutation of any static data. See https://godbolt.org/z/eGhx9b for example.

I've got an implementation of this that I plan on merging in along with a batch of other changes.

**Pull request closed without merging**